cmd/client-s3.go

```diff
@@ -2120,57 +2131,61 @@ func (c *S3Client) ShareUpload(isRecursive bool, expires time.Duration, contentT
 }
 
 // SetObjectLockConfig - Set object lock configurataion of bucket.
-func (c *S3Client) SetObjectLockConfig(mode *minio.RetentionMode, validity *uint, unit *minio.ValidityUnit) *probe.Error {
+func (c *S3Client) SetObjectLockConfig(mode minio.RetentionMode, validity uint64, unit minio.ValidityUnit) *probe.Error {
 	bucket, _ := c.url2BucketAndObject()
 
-	err := c.api.SetBucketObjectLockConfig(bucket, mode, validity, unit)
-	if err != nil {
-		return probe.NewError(err)
+	// FIXME: This is too ugly, fix minio-go
+	vuint := (uint)(validity)
+	e := c.api.SetBucketObjectLockConfig(bucket, &mode, &vuint, &unit)
+	if e != nil {
+		return probe.NewError(e).Trace(c.GetURL().String())
 	}
 
 	return nil
 }
 
 // PutObjectRetention - Set object retention for a given object.
-func (c *S3Client) PutObjectRetention(mode *minio.RetentionMode, retainUntilDate *time.Time, bypassGovernance bool) *probe.Error {
+func (c *S3Client) PutObjectRetention(mode minio.RetentionMode, retainUntilDate time.Time, bypassGovernance bool) *probe.Error {
 	bucket, object := c.url2BucketAndObject()
 
 	opts := minio.PutObjectRetentionOptions{
-		RetainUntilDate:  retainUntilDate,
-		Mode:             mode,
+		RetainUntilDate:  &retainUntilDate,
+		Mode:             &mode,
 		GovernanceBypass: bypassGovernance,
 	}
-	err := c.api.PutObjectRetention(bucket, object, opts)
-	if err != nil {
-		return probe.NewError(err)
+	e := c.api.PutObjectRetention(bucket, object, opts)
+	if e != nil {
+		return probe.NewError(e).Trace(c.GetURL().String())
 	}
 
 	return nil
 }
 
 // PutObjectLegalHold - Set object legal hold for a given object.
-func (c *S3Client) PutObjectLegalHold(lhold *minio.LegalHoldStatus) *probe.Error {
+func (c *S3Client) PutObjectLegalHold(lhold minio.LegalHoldStatus) *probe.Error {
 	bucket, object := c.url2BucketAndObject()
 	opts := minio.PutObjectLegalHoldOptions{
-		Status: lhold,
+		Status: &lhold,
 	}
-	err := c.api.PutObjectLegalHold(bucket, object, opts)
-	if err != nil {
-		return probe.NewError(err)
+	e := c.api.PutObjectLegalHold(bucket, object, opts)
+	if e != nil {
+		return probe.NewError(e).Trace(c.GetURL().String())
 	}
 	return nil
 }
 
 // GetObjectLockConfig - Get object lock configuration of bucket.
-func (c *S3Client) GetObjectLockConfig() (mode *minio.RetentionMode, validity *uint, unit *minio.ValidityUnit, perr *probe.Error) {
+func (c *S3Client) GetObjectLockConfig() (minio.RetentionMode, uint64, minio.ValidityUnit, *probe.Error) {
 	bucket, _ := c.url2BucketAndObject()
 
-	mode, validity, unit, err := c.api.GetBucketObjectLockConfig(bucket)
-	if err != nil {
-		return nil, nil, nil, probe.NewError(err)
+	mode, validity, unit, e := c.api.GetBucketObjectLockConfig(bucket)
+	if e != nil {
+		return "", 0, "", probe.NewError(e).Trace(c.GetURL().String())
 	}
 
-	return mode, validity, unit, nil
+	// FIXME: this is too ugly, fix minio-go
+	vuint64 := uint64(*validity)
```

Comment:
Should be something like 

```
var vuint64 uint64
	if validity != nil {
		vuint64 = uint64(*validity)
	}
	var retentionMode minio.RetentionMode
	if mode != nil {
		retentionMode = *mode
	}

	var validityUnit minio.ValidityUnit
	if unit != nil {
		validityUnit = *unit
	}
	return retentionMode, vuint64, validityUnit, nil
```
As `minio-go` returns `nil` when `objectlockconfig` is not set

Reply:
fixed